src/device/sound.rs

```diff
@@ -0,0 +1,1563 @@
+    VirtioSndPcmFmtS20,          /* 20 / 32 bits */
+    VirtioSndPcmFmtU20,          /* 20 / 32 bits */
+    VirtioSndPcmFmtS24,          /* 24 / 32 bits */
+    VirtioSndPcmFmtU24,          /* 24 / 32 bits */
+    VirtioSndPcmFmtS32,          /* 32 / 32 bits */
+    VirtioSndPcmFmtU32,          /* 32 / 32 bits */
+    VirtioSndPcmFmtFloat,        /* 32 / 32 bits */
+    VirtioSndPcmFmtFloat64,      /* 64 / 64 bits */
+    /* digital formats (width / physical width) */
+    VirtioSndPcmFmtDsdU8,          /* 8 / 8 bits */
+    VirtioSndPcmFmtDsdU16,         /* 16 / 16 bits */
+    VirtioSndPcmFmtDsdU32,         /* 32 / 32 bits */
+    VirtioSndPcmFmtIec958Subframe, /* 32 / 32 bits */
+}
+
+impl From<PcmSampleFormat> for u64 {
+    fn from(value: PcmSampleFormat) -> Self {
+        value as _
+    }
+}
+
+enum PcmFrameRate {
+    VirtioSndPcmRate5512 = 0,
+    VirtioSndPcmRate8000,
+    VirtioSndPcmRate11025,
+    VirtioSndPcmRate16000,
+    VirtioSndPcmRate22050,
+    VirtioSndPcmRate32000,
+    VirtioSndPcmRate44100,
+    VirtioSndPcmRate48000,
+    VirtioSndPcmRate64000,
+    VirtioSndPcmRate88200,
+    VirtioSndPcmRate96000,
+    VirtioSndPcmRate176400,
+    VirtioSndPcmRate192000,
+    VirtioSndPcmRate384000,
+}
+
+impl Into<u64> for PcmFrameRate {
+    fn into(self) -> u64 {
+        1 << self as usize
+    }
+}
+
+/// PCM information.
+#[repr(C)]
+#[derive(AsBytes, FromBytes, FromZeroes)]
+pub struct VirtIOSndPcmInfo {
+    hdr: VirtIOSndInfo,
+    features: u32, /* 1 << VIRTIO_SND_PCM_F_XXX */
+    formats: u64,  /* 1 << VIRTIO_SND_PCM_FMT_XXX */
+    rate: u64,     /* 1 << VIRTIO_SND_PCM_RATE_XXX */
+    /// indicates the direction of data flow (VIRTIO_SND_D_*)
+    direction: u8,
+    /// indicates a minimum number of supported channels
+    channels_min: u8,
+    /// indicates a maximum number of supported channels
+    channels_max: u8,
+    _padding: [u8; 5],
+}
+
+impl Display for VirtIOSndPcmInfo {
+    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
+        let mut features = String::new();
+        bitflags::parser::to_writer(&PcmFeatures::from(self.features), &mut features).unwrap();
+        let mut rates = String::new();
+        let _ = bitflags::parser::to_writer(&PcmRate::from(self.rate), &mut rates).unwrap();
+        let mut formats = String::new();
+        let _ = bitflags::parser::to_writer(&PcmFormats::from(self.formats), &mut formats).unwrap();
+        let direction = if self.direction == VIRTIO_SND_D_INPUT {
+            "INPUT"
+        } else {
+            "OUTPUT"
+        };
+        write!(
+            f,
+            "features: {}, rates: {}, formats: {}, direction: {}",
+            features, rates, formats, direction
+        )
+    }
+}
+
+struct PcmParameters {
+    setup: bool,
+    buffer_bytes: u32,
+    period_bytes: u32,
+    features: PcmFeatures,
+    channels: u8,
+    format: PcmFormats,
+    rate: PcmRate,
+}
+
+impl Default for PcmParameters {
```

Comment:
Can't this be derived automatically?

Reply:
Fixed, now this is derived automatically.